### ISSUE TITLE
Fish master doesn't seem to echo characters back on serial login

### ISSUE THREAD

**Issue by Depau** [fish-shell/fish-shell#7709]
Affects current master 056153e, works fine with latest stable - potentially a regression

After login (if set as the default shell) or when logging into bash and then running fish, it won't display the prompt or echo characters. I eventually realized it still accepts commands, and that `reset` seems to fix it

It happens with HOME set to a temp directory as well.

https://asciinema.org/a/S2WVPi9kOqZC2cUbZJm0BDdzP

Not shown in the asciirec but tested: running in this state that prints something works, including `echo` and `stty`.

`stty -a` while in this state:

```
speed 115200 baud; rows 0; columns 0; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 hupcl -cstopb cread clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl -ixon -ixoff
-iuclc -ixany -imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
-echoctl echoke -flusho -extproc
```

after reset
```
lspeed 115200 baud; rows 24; columns 80; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 hupcl -cstopb cread -clocal -crtscts
-ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff
-iuclc -ixany imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc
```

It only seems to occur over serial. On virtual terminals and graphical terminal emulators it works normally.

To ensure this wasn't related to the UART-to-WebSocket device I talked about in #7708, I also tested this by connecting two UART adapters together, running `getty` on one and `tio` on the other - this is what you see in the asciinema rec.

**Comment by Depau** [comment]
This issue seems to persist until the terminal is reset, and only with fish - bash works fine: https://asciinema.org/a/g5AhRU2ofhNb1qgQSmcgLDt4x

Not shown but tested working: tset also fixes it. This can be run even on bash, when running fish subsequently it seems to work just fine.

**Comment by faho** [comment]
Ah, okay.

The only relevant difference here seems to be that rows/columns is unset (e.g. "echoctl" is most likely not the culprit). Typically fish just asks the terminal via ioctl(TIOCGWINSZ), I'm guessing in your case that doesn't work.

I *think* we might have stopped falling back to 80x24 by accident.

**Comment by Depau** [comment]
Yep, that seems to be the case (newlines and `# comments` mine):

```
# terminal "stuck"

# Run `stty size`
0 0

# Run `resize`
COLUMNS=239;
LINES=54;
export COLUMNS LINES;

depau@frigo ~/B/t/c/ttyc (main)> # prompt is now shown correctly
```

**Comment by faho** [comment]
Okay, in c7160d7cb4970c2a03df34547f357721cb5e88db we indeed lost the default if the terminal answers with 0.

Lemme just default it to 80x24 again.

**Comment by faho** [comment]
Should be fixed by a503371 - if we get a 0 for either dimension, we assume the default instead.

**Comment by Depau** [comment]
I can confirm it works, thank you.
